library/general/uniqueingresshost/template.yaml

```diff
@@ -0,0 +1,24 @@
+apiVersion: templates.gatekeeper.sh/v1beta1
+kind: ConstraintTemplate
+metadata:
+  name: k8suniqueingresshost
+spec:
+  crd:
+    spec:
+      names:
+        kind: K8sUniqueIngressHost
+        listKind: K8sUniqueIngressHostList
+        plural: k8suniqueingresshost
+        singular: k8suniqueingresshost
+  targets:
+    - target: admission.k8s.gatekeeper.sh
+      rego: |
+        package k8suniqueingresshost
+       
+        violation[{"msg": msg}] {
+          input.review.kind.kind == "Ingress"              
+          host := input.review.object.spec.rules[_].host
+          ingresses := data.inventory.namespace[_][_]["Ingress"]
+          ingresses[_].spec.rules[_].host == host
+          msg := sprintf("ingress host conflicts with an existing ingress <%v>", [host])
```

Comment:
This will always fail on audit (assuming at least one host-based ingress) and update unless you filter out duplicate ingresses (e.g. group + version + kind + namespace + name are identical)

Reply:
Something like this would do it:

```
some ns, name
ingresses := data.inventory.namespace[ns][_]["Ingress"]
ingresses[name].spec.rules[_].host == host
name != input.review.name
ns != input.review.namespace
```

If we're still using OPA < v0.11 then remove `some ns, name`.

Reply:
I think I've noticed some inconsistency in whether `input.review.name` gets defined or not. It may be worth getting `input.review.name` by default and falling back to `input.review.object.metadata.name` for safety.

Reply:
Did we validate that input.review.name gets set for Ingresses?

Reply:
From the tests I have done in my own test cluster, I did not see `name` in `input.review` for ingress.